include/sys/mpsc_pbuf.h

```diff
@@ -0,0 +1,237 @@
+
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef ZEPHYR_INCLUDE_SYS_MPSC_PBUF_H_
+#define ZEPHYR_INCLUDE_SYS_MPSC_PBUF_H_
+
+#include <kernel.h>
+#include <sys/mpsc_packet.h>
+#include <string.h>
+#include <stdint.h>
+#include <stdbool.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Multi producer, single consumer packet buffer API
+ * @defgroup mpsc_buf MPSC (Multi producer, single consumer) packet buffer API
+ * @ingroup kernel_apis
+ * @{
+ */
+
+/*
+ * Multi producer, single consumer packet buffer allows to allocate variable
+ * length consecutive space for storing a packet. When space is allocated
+ * it can be filled by the user (except for the first 2 bits) and when packet
+ * is ready it is commited. It is allowed to allocate another packet before
+ * commiting the previous one.
+ *
+ * If buffer is full and packet cannot be allocated then null is returned unless
+ * overwrite mode is selected. In that mode, oldest entry are dropped (user is
+ * notified) until allocation succeeds. It can happen that candidate for
+ * dropping is currently being claimed. In that case, it is ommited and next
+ * packet is dropped and claimed packet is marked as invalid when freeing.
+ *
+ * Reading packets is performed in two steps. First packet is claimed. Claiming
+ * returns pointer to the packet within the buffer. Packet is freed when no
+ * longer in use.
+ */
+
+/**@defgroup MPSC_PBUF_FLAGS MPSC packet buffer flags
+ * @{ */
+
+/** @brief Flag indicating that buffer size is power of 2.
+ *
+ * When buffer size is power of 2 then optimizations are applied.
+ */
+#define MPSC_PBUF_SIZE_POW2 BIT(0)
+
+/** @brief Flag indicating buffer full policy.
+ *
+ * If flag is set then when allocating from a full buffer oldest packets are
+ * dropped. When flag is not set then allocation returns null.
+ */
+#define MPSC_PBUF_MODE_OVERWRITE BIT(1)
+
+/**@} */
+
+/* Forward declaration */
+struct mpsc_pbuf_buffer;
+
+/** @brief Callback prototype for getting length of a packet.
+ *
+ * @param packet User packet.
+ *
+ * @return Size of the packet in 32 bit words.
+ */
+typedef uint32_t (*mpsc_pbuf_get_wlen)(union mpsc_pbuf_generic *packet);
+
+/** @brief Callback called when packet is dropped.
+ *
+ * @param buffer Packet buffer.
+ *
+ * @param packet Packet that is being dropped.
+ */
+typedef void (*mpsc_pbuf_notify_drop)(struct mpsc_pbuf_buffer *buffer,
+					union mpsc_pbuf_generic *packet);
+
+/** @brief MPSC packet buffer structure. */
+struct mpsc_pbuf_buffer {
+	/* Temporary write index. */
+	uint32_t tmp_wr_idx;
+
+	/* Write index. */
+	uint32_t wr_idx;
+
+	/* Temporary read index. */
+	uint32_t tmp_rd_idx;
+
+	/* Read index. */
+	uint32_t rd_idx;
+
+	uint32_t flags;
```

Comment:
All fields should have a document comment, or you need to hide them separately. The generated documentation will look a bit unfriendly to the user reading the docs now.